### ISSUE TITLE
BUG: Fail to call `BSpline` after unpickling with `mmap_mode="r"`

### ISSUE THREAD

**Issue by FBruzzesi** [scipy/scipy#22143]
### Describe your issue.

`BSpline` fails to run after unpickling with `mmap_mode="r"`.
This is a regression in the release candidate `scipy==1.15.0rc1`, and does not happen in `scipy==1.14.0`, and it was originally spotted in scikit-lego where we run test suite against release candidates as well.

### Reproducing Code Example

```python
import tempfile
import joblib
from pathlib import Path
from scipy.interpolate import BSpline

# from the example in the docs
k = 2
t = [0, 1, 2, 3, 4, 5, 6]
c = [-1, 2, 0, -1]
estimator = BSpline(t=t, c=c, k=k)

temp_folder = tempfile.mkdtemp(prefix="tmp")
filename = Path(temp_folder) / "bspline.pkl"
joblib.dump(estimator, filename)

unpickled_estimator = joblib.load(filename, mmap_mode="r")
unpickled_estimator([1,2,3])  # <--- this call will fail
```

<details>
<summary> Originally code that spotted the issue </summary>

```py
from sklearn.preprocessing import SplineTransformer
from sklearn.utils.estimator_checks import check_estimators_pickle

check_estimators_pickle(
    name="hello",
    estimator_orig=SplineTransformer(),
    readonly_memmap=True,
)
```
</details>

### Error message

```shell
Traceback (most recent call last):
  File "/home/fbruzzesi/open-source/scikit-lego/t.py", line 21, in <module>
    unpickled_estimator([1,2,3])
  File "/home/fbruzzesi/open-source/scikit-lego/.venv-pre/lib/python3.10/site-packages/scipy/interpolate/_bsplines.py", line 523, in __call__
    _dierckx.evaluate_spline(self.t, cc.reshape(cc.shape[0], -1),
ValueError: Expected a 1-dim C contiguous array  of dtype = 12( got 12 )
```

### SciPy/NumPy/Python version and system information

```shell
1.15.0rc1 2.2.0 sys.version_info(major=3, minor=10, micro=12, releaselevel='final', serial=0)
/home/fbruzzesi/open-source/scikit-lego/.venv-pre/lib/python3.10/site-packages/scipy/__config__.py:154: UserWarning: Install `pyyaml` for better output
  warnings.warn("Install `pyyaml` for better output", stacklevel=1)
{
  "Compilers": {
    "c": {
      "name": "gcc",
      "linker": "ld.bfd",
      "version": "10.2.1",
      "commands": "cc"
    },
    "cython": {
      "name": "cython",
      "linker": "cython",
      "version": "3.0.11",
      "commands": "cython"
    },
    "c++": {
      "name": "gcc",
      "linker": "ld.bfd",
      "version": "10.2.1",
      "commands": "c++"
    },
    "fortran": {
      "name": "gcc",
      "linker": "ld.bfd",
      "version": "10.2.1",
      "commands": "gfortran"
    },
    "pythran": {
      "version": "0.17.0",
      "include directory": "../../tmp/pip-build-env-pxy77qn_/overlay/lib/python3.10/site-packages/pythran"
    }
  },
  "Machine Information": {
    "host": {
      "cpu": "x86_64",
      "family": "x86_64",
      "endian": "little",
      "system": "linux"
    },
    "build": {
      "cpu": "x86_64",
      "family": "x86_64",
      "endian": "little",
      "system": "linux"
    },
    "cross-compiled": false
  },
  "Build Dependencies": {
    "blas": {
      "name": "scipy-openblas",
      "found": true,
      "version": "0.3.28",
      "detection method": "pkgconfig",
      "include directory": "/opt/_internal/cpython-3.10.15/lib/python3.10/site-packages/scipy_openblas32/include",
      "lib directory": "/opt/_internal/cpython-3.10.15/lib/python3.10/site-packages/scipy_openblas32/lib",
      "openblas configuration": "OpenBLAS 0.3.28 DYNAMIC_ARCH NO_AFFINITY Haswell MAX_THREADS=64",
      "pc file directory": "/project"
    },
    "lapack": {
      "name": "scipy-openblas",
      "found": true,
      "version": "0.3.28",
      "detection method": "pkgconfig",
      "include directory": "/opt/_internal/cpython-3.10.15/lib/python3.10/site-packages/scipy_openblas32/include",
      "lib directory": "/opt/_internal/cpython-3.10.15/lib/python3.10/site-packages/scipy_openblas32/lib",
      "openblas configuration": "OpenBLAS 0.3.28 DYNAMIC_ARCH NO_AFFINITY Haswell MAX_THREADS=64",
      "pc file directory": "/project"
    },
    "pybind11": {
      "name": "pybind11",
      "version": "2.13.6",
      "detection method": "config-tool",
      "include directory": "unknown"
    }
  },
  "Python Information": {
    "path": "/opt/python/cp310-cp310/bin/python",
    "version": "3.10"
  }
}
```

**Comment by ev-br** [comment]
Interesting. Would be great to distill the MRE to just scipy without scikit-learn. 
Also, would be great to check if the issue is the same in scipy==1.4.1, as there was a change in the internals in 1.15.x 


```
In [1]: import pickle

In [2]: from scipy.interpolate import make_interp_spline

In [3]: import numpy as np

In [4]: x = np.arange(5)

In [5]: spl = make_interp_spline(x, x**3, k=3)

In [6]: pkl = pickle.dumps(spl)

In [7]: u = pickle.loads(pkl)

In [8]: spl(3)
Out[8]: array(27.)

In [9]: u(3)
Out[9]: array(27.)

In [10]: import scipy

In [11]: scipy.__version__
Out[11]: '1.16.0.dev0+git20241030.bcd75c3'

```

**Comment by ev-br** [comment]
Ah, so it's a regression from https://github.com/scipy/scipy/pull/21551
As https://github.com/scipy/scipy/issues/22143#issuecomment-2556793442 seems to indicate just pickling a BSpline instance is still OK in scipy >= 1.15. Would be super helpful to distill it further avoid `from sklearn* import`s, if possible.

**Comment by lesteve** [comment]
Here is a way to reproduce without joblib:
```py
import numpy as np
from scipy.interpolate import BSpline

k = 2
t = [0, 1, 2, 3, 4, 5, 6]
c = [-1, 2, 0, -1]

t_mm = np.memmap('knots.dat', dtype='float64', mode='w+', shape=(7,))
t_mm[:] = t
t_mm.flush()

print('constructing BSpline from mmap')
spline = BSpline(t=t_mm, c=c, k=k)
print(f"{spline.t=}")
print(f"{spline.t.base=}")
spline(0)  # works

print('setting attribute directly')
spline.t = t_mm
print(f"{spline.t=}")
print(f"{spline.t.base=}")
spline(0)  # ValueError: Expected a 1-dim C contiguous array  of dtype = 12( got 12 )

```

Output:
```
constructing BSpline from mmap
spline.t=array([0., 1., 2., 3., 4., 5., 6.])
spline.t.base=memmap([0., 1., 2., 3., 4., 5., 6.])
setting attribute directly
spline.t=memmap([0., 1., 2., 3., 4., 5., 6.])
spline.t.base=<mmap.mmap closed=False, access=ACCESS_WRITE, length=56, pos=0, offset=0>
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
File /tmp/test.py:24
     22 print(f"{spline.t=}")
     23 print(f"{spline.t.base=}")
---> 24 spline(0)  # ValueError: Expected a 1-dim C contiguous array  of dtype = 12( got 12 )

File ~/micromamba/envs/scikit-learn-dev-dependencies/lib/python3.13/site-packages/scipy/interpolate/_bsplines.py:523, in BSpline.__call__(self, x, nu, extrapolate)
    520 if self.c.ndim == 1 and self.c.dtype.kind == 'c':
    521     cc = cc.reshape(self.c.shape[0], 2)
--> 523 _dierckx.evaluate_spline(self.t, cc.reshape(cc.shape[0], -1),
    524                       self.k, x, nu, extrapolate, out.view(float))
    526 out = out.reshape(x_shape + self.c.shape[1:])
    527 if self.axis != 0:
    528     # transpose to move the calculated values to the interpolation axis

ValueError: Expected a 1-dim C contiguous array  of dtype = 12( got 12 )
```

I would agree this doesn't seem natural to set the attribute to a memmap after constructing it. But this seems similar to what joblib is doing so this will affect scikit-learn if `BSpline` are used in parallel processing, as I guess could happen if you do a cross-validation where the scikit-learn estimator use a `SplineTransformer`.

**Comment by ev-br** [comment]
Thanks! So the issue is clear: dropping a breakpoint into `BSpline.__call__`  I see

```
(Pdb) p self.t
memmap([0., 1., 2., 3., 4., 5., 6.])
(Pdb) p type(self.t)
<class 'numpy.memmap'>
```

and then in `_dierckx.evaluate_spline` call, there is `PyArray_CheckExact`: https://github.com/scipy/scipy/blob/main/scipy/interpolate/src/_dierckxmodule.cc#L16
which emits the observed error message.

In scipy 1.14.1 and below, this evaluation went through cython typed memoryviews. Have to admit I don't know much about numpy memmaps, do they support buffer protocol? That would explain why it worked before and does not now.

BTW, note that `cc` array is derived from `self.c`. In the MRE, it is a regular array not a memmap. So if the joblib code is meant to turn all arrays into memmaps, this one has been turned into a regular array already. This code has seen several changes in the 1.15 cycle.

----------------------------------------------------

Now, what to do about it. The code in the new `_dierckx` module basically does this: check an input is a numpy arrays, grab the data pointer and shapes, and pass it to low-level routines for heavy lifting. It certainly needs to have a pointer to a C contiguous array.
Is there a way to grab this info from a memmapped array---do they even have a pointer to data?

**Comment by lesteve** [comment]
I am definitely not an expert on `np.memmap` in general but it looks like cython memoryview supports it:

```
In [1]: %load_ext cython

In [2]: %%cython
   ...: import numpy as np
   ...: 
   ...: def func(const double[:] arr):
   ...:     return np.asarray(arr).sum()
   ...: 
   ...: 

In [3]: t_mm = np.memmap('knots.dat', dtype='float64', mode='w+', shape=(7,))
   ...: t_mm[:] = np.ones(7)
   ...: func(t_mm)
   ...: 
Out[3]: np.float64(7.0)
```

For the context we actually reported and fixed an issue in Scipy that caused us to xfail a test in https://github.com/scipy/scipy/pull/18153 but did not un-xfail our test otherwise we would have caught this a while ago when the change was done in scipy :sweat: ...

**Comment by ev-br** [comment]
@tylerjereddy: the 1.15 milestone is mainly FYI here. It does not look like a fix here is going to materialize in the next couple of days, and even if it does, it'll involve C code, and it will make C code which is new in 1.15 more permissive. 

So my suggestion to you is to revert gh-21551 in the maintenance/1.15.x branch. That'll make 1.15.x branch use stuff which caused zero problems for a long time, and we'll be able to investigate a proper fix in the 1.16.x timeframe. 

@lesteve am wondering if scikit-learn is testing scipy main or nightlies somewhere on the CI?

**Comment by tylerjereddy** [comment]
@ev-br thanks for the heads up, I've added a link to your comment in the TODO list in gh-22150.

**Comment by lesteve** [comment]
I have opened https://github.com/scipy/scipy/pull/22158 which is a tentative fix by accepting ndarray subclasses in the check (`PyArray_Check` rather than `PyArray_CheckExact`)

**Comment by ev-br** [comment]
Just to confirm: this fixes the reported OP with joblib?

**Comment by lesteve** [comment]
> Just to confirm: this fixes the reported OP with joblib?

Yes

> @lesteve am wondering if scikit-learn is testing scipy main or nightlies somewhere on the CI?

Forgot to answer this one: we do test against scipy nightlies and we should have detected the regression sooner in principle. Unfortunately we xfailed this particular test for historical reason and never un-xfailed it back after https://github.com/scipy/scipy/pull/18153 was merged ... I have opened https://github.com/scikit-learn/scikit-learn/pull/30515 to un-xfail this particular test.
